clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java

```diff
@@ -341,12 +353,24 @@ public void onCompletion(Optional<Exception> exception) {
           if (!exceptionToSave.isPresent()) {
             finalRequestCompletionFuture.complete(null);
           } else {
+            boolean shouldCompleteRequestFuture = false;
+            if (scheduledRetryTask != null && isExceptionCausedByTooManyRequests(exceptionToSave.get())) {
+              // Check if the exception is 429, if so cancel the retry if the retry task exists
```

Comment:
as the retry task checks `savedException`, for the defensive code to be useful, I think we need to do this after 
`if (!savedException.compareAndSet(null, exceptionToSave.get())) {` block.

Reply:
Good catch, will swap it for better coverage but just want to be clear that it's still not bullet proof and we try to cancel/avoid the retry as much as we can. Since the retry could be triggered already before we reach this block of code in the original request.

Reply:
yes, agree with this.